prov/opx/src/fi_opx_init.c

```diff
@@ -0,0 +1,481 @@
+		fi->tx_attr->op_flags = hints->tx_attr->op_flags;
+	} else if (hints && hints->caps) {
+		fi->tx_attr->caps = hints->caps;
+	}
+
+	memcpy(fi->rx_attr, fi_opx_global.default_rx_attr, sizeof(*fi->rx_attr));
+	if (hints && hints->rx_attr) {
+
+		/*
+		 * man/fi_endpoint.3
+		 *
+		 *   fi_rx_attr::caps
+		 *
+		 *   "... If the caps field is 0 on input to fi_getinfo(3), the
+		 *   caps value from the fi_info structure will be used."
+		 */
+		if (hints->rx_attr->caps) {
+			fi->rx_attr->caps = hints->rx_attr->caps;
+		}
+
+		/* adjust parameters down from what requested if required */
+		fi->rx_attr->op_flags = hints->rx_attr->op_flags;
+		if (hints->rx_attr->total_buffered_recv > 0 &&
+			hints->rx_attr->total_buffered_recv < fi_opx_global.default_rx_attr->total_buffered_recv)
+				fi->rx_attr->total_buffered_recv = hints->rx_attr->total_buffered_recv;
+	} else if (hints && hints->caps) {
+		fi->rx_attr->caps = hints->caps;
+	}
+
+	caps = fi->caps | fi->tx_attr->caps | fi->rx_attr->caps;
+
+	/*
+	 * man/fi_domain.3
+	 *
+	 * On input to fi_getinfo, a user may set this (fi_domain_attr::domain)
+	 * to an opened domain instance to restrict output to the given domain.
+	 * On output from fi_getinfo, if no domain was specified, but the user
+	 * has an opened instance of the named domain, this (fi_domain_attr::domain)
+	 * will reference the first opened instance. If no instance has been
+	 * opened, this field will be NULL.
+	 */
+
+	ret = fi_opx_choose_domain(caps, fi->domain_attr,
+		(hints)?(hints->domain_attr):NULL);
+	if (ret) {
+		FI_LOG(fi_opx_global.prov, FI_LOG_DEBUG, FI_LOG_FABRIC,
+				"cannot find appropriate domain\n");
+		goto err;
+	}
+
+	memcpy(fi->ep_attr, fi_opx_global.default_ep_attr, sizeof(*fi->ep_attr));
+	if (hints && hints->ep_attr) {
+		/* adjust parameters down from what requested if required */
+		fi->ep_attr->type	= hints->ep_attr->type;
+		if (hints->ep_attr->max_msg_size > 0 &&
+			hints->ep_attr->max_msg_size <= fi_opx_global.default_ep_attr->max_msg_size)
+				fi->ep_attr->max_msg_size = hints->ep_attr->max_msg_size;
+
+		if (0 != hints->ep_attr->tx_ctx_cnt && hints->ep_attr->tx_ctx_cnt <= fi->ep_attr->tx_ctx_cnt)
+			fi->ep_attr->tx_ctx_cnt = hints->ep_attr->tx_ctx_cnt;	/* TODO - check */
+
+		if (0 != hints->ep_attr->rx_ctx_cnt && hints->ep_attr->rx_ctx_cnt <= fi->ep_attr->rx_ctx_cnt)
+			fi->ep_attr->rx_ctx_cnt = hints->ep_attr->rx_ctx_cnt;	/* TODO - check */
+	}
+
+	return 0;
+err:
+	if (fi->domain_attr->name) free(fi->domain_attr->name);
+	if (fi->fabric_attr->name) free(fi->fabric_attr->name);
+	if (fi->fabric_attr->prov_name) free(fi->fabric_attr->prov_name);
+	errno = FI_ENODATA;
+	return -errno;
+}
+
+struct fi_opx_global_data fi_opx_global;
+
+static int fi_opx_getinfo(uint32_t version, const char *node,
+		const char *service, uint64_t flags,
+		const struct fi_info *hints, struct fi_info **info)
+{
+	int ret;
+	struct fi_info *fi;
+
+	if (!fi_opx_count) {
+		errno = FI_ENODATA;
+		return -errno;
```

Comment:
On error should set `*info = NULL;`. This is probably the reason for the different behavior of the `fi_info` test.

Reply:
@j-xiong 

Oof. I thought of that, but forcing it to NULL has no effect. I made the following changes:

```
diff --git a/prov/opx/src/fi_opx_init.c b/prov/opx/src/fi_opx_init.c
index 5cadb5dcb..c28d82478 100644
--- a/prov/opx/src/fi_opx_init.c
+++ b/prov/opx/src/fi_opx_init.c
@@ -361,6 +361,8 @@ static int fi_opx_getinfo(uint32_t version, const char *node,
    int ret;
    struct fi_info *fi;
 
+   *info = NULL;
+
    if (!fi_opx_count) {
        errno = FI_ENODATA;
        return -errno;
diff --git a/src/fabric.c b/src/fabric.c
index 8347dc3f7..8970ec1e8 100644
--- a/src/fabric.c
+++ b/src/fabric.c
@@ -1120,6 +1120,9 @@ int DEFAULT_SYMVER_PRE(fi_getinfo)(uint32_t version, const char *node,
        if (!prov->provider || !prov->provider->getinfo)
            continue;
 
+       //if (strcmp(prov->provider->name, "opx") == 0) continue; 
+       fprintf(stderr,"MHEINZ: %s entry %p\n",prov->provider->name, prov);
+
        if (prov->hidden && !(flags & OFI_GETINFO_HIDDEN))
            continue;
 
@@ -1139,6 +1142,8 @@ int DEFAULT_SYMVER_PRE(fi_getinfo)(uint32_t version, const char *node,
        cur = NULL;
        ret = prov->provider->getinfo(version, node, service, flags,
                          hints, &cur);
+       fprintf(stderr,"MHEINZ: %s returned %p / %d\n",prov->provider->name, cur, ret);
+
        if (ret) {
            level = ((hints && hints->fabric_attr &&
                  hints->fabric_attr->prov_name) ?
[```
but behavior is still broken:

```
[awbp023:~/work/libfabric-internal](rc-opx-1.15.0)$ ~/OPX-Build/ofi-opx/bin/fi_info -f -p sockets 2>&1 | tee a
MHEINZ: opx entry 0x184f820
MHEINZ: opx returned (nil) / -61
MHEINZ: psm2 entry 0x184f870
MHEINZ: psm2 returned (nil) / -61
MHEINZ: usnic entry 0x184f910
MHEINZ: usnic returned (nil) / -61
MHEINZ: verbs entry 0x184fa00
MHEINZ: verbs returned (nil) / -61
MHEINZ: psm3 entry 0x184faa0
MHEINZ: psm3 returned (nil) / -61
MHEINZ: ofi_rxm entry 0x184faf0
MHEINZ: ofi_rxm returned (nil) / -61
MHEINZ: ofi_rxd entry 0x184fb40
MHEINZ: ofi_rxd returned (nil) / -61
MHEINZ: shm entry 0x184fb90
MHEINZ: shm returned (nil) / -61
MHEINZ: udp entry 0x184fbe0
MHEINZ: udp returned (nil) / -61
MHEINZ: tcp entry 0x184fc30
MHEINZ: tcp returned (nil) / -61
MHEINZ: sockets entry 0x184fc80
MHEINZ: sockets returned (nil) / -61
MHEINZ: ofi_mrail entry 0x1862f90
MHEINZ: ofi_mrail returned (nil) / -61
fi_getinfo: -61
```

It's definitely something to do with calling fi_opx_getinfo() though - adding a check to skip over OPX right before my first MHEINZ message fixes the problem.

Reply:
Btw, you can declare a static struct fi_info in the source, link that into a util_prov structure, and use the generic util_getinfo path.  It'll likely save you a lot of code duplication and ensure correct behavior, including API versioning support (for apps compiled against older versions of libfabric).

So far, I haven't found anything off in the getinfo path that would result in the behavior differences, but I'll keep looking.